### ISSUE TITLE
Provide copy progress feedback

### ISSUE THREAD

**Issue by danielburrell** [carvel-dev/imgpkg#96]
**Describe the problem/challenge you have**
As a user, I am pushing images to a docker registry as part of an airgap deployment. Sometimes this takes a long time and I'm left wondering if progress is being made. I'm not calling imgpkg directly, it's being wrapped.

**Describe the solution you'd like**
Some sort of parsable output indicating the progress being made as if it were an uploading data file. "305k/3000k - 10% ETA 30 seconds" this information is considered useful because it shows how much has been uploaded, how much must be uploaded, expresses that as a percentage for simplicity, and then shows the expected time it will complete.

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

**Comment by pivotaljohn** [comment]
Complete agreement, @danielburrell.

It'll be a minute...

To do this, we're preparing a contribution to our upstream dependency (we use GGCR for all our direct registry interactions): https://github.com/google/go-containerregistry/issues/952

With progress reports coming in from GGCR, `imgpkg` can then provide that up-to-date status. 👍🏼 

Watch this space for updates.

**Comment by DennisDenuto** [comment]
There has been a [PR submitted upstream](https://github.com/google/go-containerregistry/pull/967) to go-container-registry that should allow improving the status feedback from imgpkg.

Once that PR has been merged in, and a new version of go-container-registry has been released, we should be able to work on this.

**Comment by aaronshurley** [comment]
Before starting this work, let's add some acceptance criteria here so that we consider what the UX will be.

**Comment by StevenLocke** [comment]
### Current `imgpkg copy` output during a copy of cf for k8s:

```
"copy | importing 45 images..."

"copy | importing index.docker.io/cloudfoundry/eirini-event-reporter-cf-for-k8s@sha256:1798ea550cfb7b1457698dc9a9766ad5333da05c1e0cd29e748b7db34e648e1b -> index.docker.io/k8slt/cf-for-k8s-from-tar-2@sha256:1798ea550cfb7b1457698dc9a9766ad5333da05c1e0cd29e748b7db34e648e1b..."

<the above will be printed as each image is verified and tagged>

"copy | imported 45 images"
"Succeeded"
```

### Progress Bar

Showing bytes copied, total bytes to copy, a percentage of those two, and an estimate for time remaining.
```
"copy | importing 45 images"

[================>_______________________________] (45% ETA: 5m10s)
55555 Kb / 1234567 Kb (100 Kb/s)

"Succeeded"
```

Notes:
* There are libraries that could help us render a progress bar, and hopefully the time remaining as well.
* This removes the "a -> b" text, showing the old and new registries. Is that valuable to maintain?
* Nice to have: a speed of transfer
* Unsure on bar position. Above/below the parseable text? May be limited by the library we choose.

**Comment by danielburrell** [comment]
These sound great if they can be achieved!

Just one thing to note, from my minimal understanding of how these console based progress bars work, they use special characters to move the cursor/redraw characters on screen, and these special characters tend to not play well with CI systems (specifically the log files). We should be careful to ensure that anything using these special characters can be switched off (or perhaps is switched off by default to maintain backward-compatibility).

As an example a similar piece of functionality in the java build tool called maven `mvn` for example would be the `--batch` mode flag which ensures that verbose download progress of dependencies is not recorded to the stdout/log files - otherwise you end up with 100,000 log lines related to a file download which makes the logs impossible to read.

That said the progress bar implementation is definitely something to aspire to for interactive shell sessions where there's a human watching. :100: 

Thanks so much for your help on this.

**Comment by joaopapereira** [comment]
Shipped with version 0.7.0
